src/trace_ng/log.zig

```diff
@@ -0,0 +1,1048 @@
+
+        pub fn logfWithFields(self: *Self, level: Level, comptime fmt: []const u8, args: anytype, fields: anytype) void {
+            if (@intFromEnum(level) > @intFromEnum(self.max_level)) {
+                // noop
+                return;
+            }
+
+            self.log_msg.?.clearAndFree();
+            const maybe_scope = if (scope) |s| s else null;
+
+            // Format fields.
+            var fmt_fields = std.ArrayList(u8).initCapacity(self.allocator, 256) catch @panic("could not initCapacity for message");
+            defer fmt_fields.deinit();
+            logfmt.fmtField(fmt_fields.writer(), fields);
+
+            // Format message.
+            var fmt_msg = std.ArrayList(u8).initCapacity(self.allocator, 256) catch @panic("could not initCapacity for message");
+            defer fmt_msg.deinit();
+            logfmt.fmtMsg(fmt_msg.writer(), fmt, args);
+
+            const log_msg = logfmt.LogMsg{
+                .level = level,
+                .maybe_scope = maybe_scope,
+                .maybe_msg = null,
+                .maybe_fields = fmt_fields.items,
+                .maybe_fmt = fmt_msg.items,
+            };
+
+            const writer = self.log_msg.?.writer();
+            logfmt.writeLog(writer, log_msg) catch @panic("Failed to write log");
+        }
+    };
+}
+
+test "trace_ng: scope switch" {
+    const StuffChild = struct {
+        const StuffChild = @This();
+        logger: ScoppedLogger(@typeName(StuffChild)),
+
+        pub fn init(logger: *const Logger) StuffChild {
+            return .{ .logger = logger.withScope(@typeName(StuffChild)) catch {
+                @panic("Init logger failed");
+            } };
+        }
+
+        pub fn deinit(self: *StuffChild) void {
+            self.logger.deinit();
+        }
+
+        pub fn doStuffDetails(self: *StuffChild) void {
+            self.logger.log(.info, "doing stuff child");
+        }
+    };
+
+    const Stuff = struct {
+        const Stuff = @This();
+        logger: ScoppedLogger(@typeName(Stuff)),
+
+        pub fn init(logger: *const Logger) Stuff {
+            return .{ .logger = logger.withScope(@typeName(Stuff)) catch @panic("Init logger failed") };
+        }
+
+        pub fn deinit(self: *Stuff) void {
+            self.logger.deinit();
+        }
+
+        pub fn doStuff(self: *Stuff) void {
+            self.logger.log(.info, "doing stuff parent");
+            const logger = self.logger.unscoped() catch @panic("Init logger failed");
+            defer logger.deinit();
+            var child = StuffChild.init(&logger);
+            defer child.deinit();
+            child.doStuffDetails();
+        }
+    };
+
+    const allocator = std.testing.allocator;
+
+    const logger = Logger.init(.{
+        .allocator = allocator,
+        .max_level = Level.info,
+        .max_buffer = 2048,
+    }) catch @panic("Logger init failed");
+    defer logger.deinit();
+
+    var stuff = Stuff.init(&logger);
+    defer stuff.deinit();
```

Comment:
I think the typical usage pattern would be that you initialize StdErrLogger somewhere near the main function of your application. You defer deinit in that scope only. Then you convert it to a Logger, which contains a pointer to StdErrLogger. The logger gets passed around everywhere by value, but it never needs to be deinited because it's just a pointer to the logger owned by main, which will be deinited when main returns.

Reply:
For scoped logger, does this not bring back the possibility for error where a scoped logger is passed and used in a different scope than was intended?

I felt having the logger be part of the state of the struct, ie

```zig

const Foo = struct {
  logger: ScopedLogger(Foo Scope)

  self.logger.info(..)
} 
```

Would ensure that logging from `Foo` would always be scoped to `Foo Scope`. 

And I am not sure that can be accomplished with the approach you described.

Reply:
Scope is not a part of the state. It's an immutable part of the type. Moving the logger to another place and giving a new scope in that place is not going to make any changes to the state of the original logger. The original logger will continue using the scope as specified in its type.